packages/protocol/src/abstracts/BaseVault.sol

```diff
@@ -986,4 +995,75 @@ abstract contract BaseVault is ERC20, SystemAccessControl, PausableVault, VaultP
       revert BaseVault__checkRebalanceFee_excessFee();
     }
   }
+
+  /// @inheritdoc IVault
+  function harvest(
+    Strategy strategy,
+    IHarvestable provider,
+    ISwapper swapper,
+    bytes memory data
+  )
+    external
+    virtual
+    hasRole(msg.sender, HARVESTER_ROLE)
+    returns (address[] memory tokens, uint256[] memory amounts)
+  {
+    //check strategy is valid for this vault
+    if (strategy == Strategy.RepayDebt) {
+      revert BaseVault__harvest_strategyNotImplemented();
+    }
+    //check provider is valid
+    if (!_isValidProvider(address(provider))) {
+      revert BaseVault__harvest_invalidProvider();
+    }
+
+    //collect rewards from provider
+    (tokens, amounts) = _harvest(provider, data);
+
+    if (tokens.length != amounts.length) {
+      revert BaseVault__harvest_invalidRewards();
+    }
+
+    for (uint256 i = 0; i < tokens.length; i++) {
+      //transfer rewards to recipient
+      IERC20(tokens[i]).safeIncreaseAllowance(msg.sender, amounts[i]);
+    }
+
+    bytes memory callData = IHarvestManager(msg.sender).completeHarvest(
+      address(this), strategy, provider, swapper, tokens, amounts
+    );
+    _completeHarvest(address(provider), callData);
+  }
+
+  /**
+   * @dev Harvests rewards from `provider`.
+   *
+   * @param provider address of provider to harvest from
+   * @param data bytes data to be used by the harvest function of the provider, specific to each provider way of harvesting
+   */
+  function _harvest(
+    IHarvestable provider,
+    bytes memory data
+  )
+    internal
+    returns (address[] memory tokens, uint256[] memory amounts)
+  {
+    bytes memory callData = abi.encodeWithSelector(provider.harvest.selector, data);
+    (bytes memory returnData) = address(provider).functionDelegateCall(
+      callData, string(abi.encodePacked("harvest", ": delegate call failed"))
+    );
+    (tokens, amounts) = abi.decode(returnData, (address[], uint256[]));
+  }
```

Comment:
There should be some type of event emitted here, essentially for all the tokens and amounts harvested and the provider.

Same applies for the upgradeable vault.